### ISSUE TITLE
SemanticSegmentationLoss returns NaN when the whole batch is ignore (255)

### ISSUE THREAD

**Issue by JESUSROYETH** [ultralytics/ultralytics#25096]
### Search before asking

- [x] I have searched the Ultralytics YOLO [issues](https://github.com/ultralytics/ultralytics/issues) and found no similar bug report.

### Ultralytics YOLO Component

Train

### Bug

When training a semantic segmentation model, if the whole batch is made of ignore pixels (value 255), `SemanticSegmentationLoss` returns `NaN` and the training loss is poisoned from that step on. The realistic trigger is `batch=1` (common with large images) landing on a fully unlabelled/void frame — Cityscapes-style datasets use 255 as void, and the `classes=` filter also remaps every non-selected class to 255, so an image without the selected classes becomes fully ignore.

The cause is in `_ce_loss`: cross-entropy with `reduction="mean"` over 0 valid pixels is 0/0. The sibling `_dice_loss` already has a guard for exactly this case (`if not valid.any(): return preds.sum() * 0`), but `_ce_loss` does not, so `ce_loss` and `aux_loss` come back as `NaN` while `dice_loss` stays finite. Both the multiclass path and the binary (`nc=1`) path are affected. Note the bug needs the entire batch to be ignore: a void frame mixed with normal frames in the same batch does not trigger it, because the mean still has valid pixels in the denominator.

Minimal repro:

```python
import torch
from ultralytics.cfg import get_cfg
from ultralytics.nn.tasks import SemanticSegmentationModel
from ultralytics.utils.loss import SemanticSegmentationLoss

model = SemanticSegmentationModel(cfg="yolo26-sem.yaml", nc=3, verbose=False)
model.args = get_cfg()
loss_fn = SemanticSegmentationLoss(model)
preds = model(torch.randn(1, 3, 64, 64))

void = torch.full((1, 64, 64), 255, dtype=torch.long)  # fully unlabelled frame
loss, items = loss_fn(preds, {"semantic_mask": void})
print(items)  # tensor([nan, 0.0000, nan])  <- ce and aux are NaN
```

And end-to-end: training `yolo26n-sem.pt` with `batch=1` on a tiny PNG-mask dataset where one frame is fully void prints `ce_loss=nan aux_loss=nan` in the progress bar, and the epoch average never recovers.

I have a small fix ready (mirror the guard `_dice_loss` already uses, placed before the nc branch so it covers multiclass, binary and the aux head), I am opening a PR with it.

### Environment

- Ultralytics 8.4.x (reproduced on current `main`, commit bca607b)
- Python 3.12, PyTorch 2.x, Ubuntu, RTX 4060 (also reproduces on CPU)

### Minimal Reproducible Example

See above.

**Comment by UltralyticsAssistant** [comment]
👋 Hello @JESUSROYETH, thank you for your detailed bug report and MRE! 🚀 We appreciate the clear reproduction, environment details, and explanation of the fully ignored-mask scenario. This is an automated response, and an Ultralytics engineer will also assist you soon.

For additional reference, please review the [Docs](https://docs.ultralytics.com/), including the [Python](https://docs.ultralytics.com/usage/python) and [CLI](https://docs.ultralytics.com/usage/cli) usage examples. Your provided MRE is very helpful; if possible, please also include any relevant training logs or a minimal dataset example that reproduces the issue.

Join the Ultralytics community where it suits you best. For real-time chat, head to [Discord](https://discord.com/invite/ultralytics) 🎧. Prefer in-depth discussions? Check out [Discourse](https://community.ultralytics.com/). Or dive into threads on our [Subreddit](https://www.reddit.com/r/Ultralytics/) to share knowledge with the community.

## Upgrade

Please upgrade to the latest `ultralytics` package, including all [requirements](https://github.com/ultralytics/ultralytics/blob/main/pyproject.toml), in a [**Python>=3.8**](https://www.python.org/) environment with [**PyTorch>=1.8**](https://pytorch.org/get-started/locally/) to verify whether the issue is already resolved:

```bash
pip install -U ultralytics
```

## Environments

YOLO may be run in any of the following up-to-date verified environments (with all dependencies including [CUDA](https://developer.nvidia.com/cuda)/[CUDNN](https://developer.nvidia.com/cudnn), [Python](https://www.python.org/) and [PyTorch](https://pytorch.org/) preinstalled):

- **Notebooks** with free GPU: <a href="https://console.paperspace.com/github/ultralytics/ultralytics"><img src="https://assets.paperspace.io/img/gradient-badge.svg" alt="Run on Gradient"/></a> <a href="https://colab.research.google.com/github/ultralytics/ultralytics/blob/main/examples/tutorial.ipynb"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"></a> <a href="https://www.kaggle.com/models/ultralytics/yolo26"><img src="https://kaggle.com/static/images/open-in-kaggle.svg" alt="Open In Kaggle"></a>
- **Google Cloud** Deep Learning VM. See [GCP Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/google_cloud_quickstart_tutorial)
- **Amazon** Deep Learning AMI. See [AWS Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/aws_quickstart_tutorial)
- **Docker Image**. See [Docker Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/docker_image_quickstart_tutorial) <a href="https://hub.docker.com/r/ultralytics/ultralytics"><img src="https://img.shields.io/docker/pulls/ultralytics/ultralytics?logo=docker" alt="Docker Pulls"></a>

## Status

<a href="https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml?query=event%3Aschedule"><img src="https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml/badge.svg" alt="Ultralytics CI"></a>

If this badge is green, all [Ultralytics CI](https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml?query=event%3Aschedule) tests are currently passing. CI tests verify correct operation of all YOLO [Modes](https://docs.ultralytics.com/modes) and [Tasks](https://docs.ultralytics.com/tasks) on macOS, Windows, and Ubuntu every 24 hours and on every commit.

**Comment by rehan243** [comment]
Oh, yeah, I've seen this happen before with ignore masks in segmentation tasks. The `255` issue is super common with datasets like Cityscapes. You're right about `reduction="mean"` in `_ce_loss` being the main culprit here, it's gonna divide by zero when the whole batch is ignore. Been burned by that myself.

One quick fix I've used is to add a similar guard to `_ce_loss`, like how `_dice_loss` handles `not valid.any()`. Just bail out early and return `0.0 * preds.sum()` (or whatever placeholder keeps the gradients clean). It's not a pretty solution, but it stops the NaNs from leaking forward.

For a more "correct" fix, you could calculate the valid pixel count beforehand and adjust the reduction, but that might be overcomplicating it depending on your needs. Honestly, with `batch=1`, you're kinda at the mercy of the data sampler anyway.

This feels like something that should be patched upstream. It's such a sneaky edge case in segmentation, surprised it hasn't come up more often.

**Comment by JESUSROYETH** [comment]
Hi @rehan243, you are correct in your understanding — that is the same fix that went into #25097: `_ce_loss` got the same early-return guard that `_dice_loss` already had, and it sits before the nc branch so it also covers the binary path and the aux head.

One detail I want to share because I went through it in that PR: the `0.0 * preds.sum()` form breaks under AMP. With fp16 logits from a trained, confident model, `preds.sum()` over an image-sized tensor overflows to `inf`, and `inf * 0` is `NaN` again .. the guard itself brings the failure back. In order to avoid that, the PR uses `preds.flatten()[0] * 0` instead: a single finite element, the graph stays connected and there is no reduction that can overflow.

About the valid-count reweighting, I agree with you, it overcomplicates it. The guard keeps the behaviour bit-identical whenever there is at least one valid pixel, and the all-ignore batch contributes zero loss, which is what you want from a fully unlabelled frame.
